Introduce aoti_call_delegate HOP

SherlockNoMad opened the pull request:
Summary:
Previously, aoti compile node is represented as a kernel-less custom op in the exported program. The node was not eager runnable, which is a common practice for numerical validation during lowering. 

I introduce a new HOP to address this. 

The schema is following
```
aoti_call_delegate(lower_moduel: AOTInductorEPModule, original_gm: fx.GraphModule, weights: List[Tensor], inputs: List[Tensor])
```

There are a few problems exposed by HOP
- AOTI expects a FX graph with weights as getattr nodes, aka stateful graph. HOP expect graph_module arguments to be stateless. Export serializer also expect a stateless graph. Currently, to make AOTI happy, I am making `original_gm` stateful, and bypassing the serialization for `original_gm`. 
- As a result, the HOP is not re-traceable, as functionalization on stateful graph module argument will fail.

Test Plan: buck2 test 'fbcode//mode/opt' fbcode//deeplearning/aot_inductor/cpu/test:cpu_lowering_utils_test

Reviewed By: zhxchen17

Differential Revision: D68359391

SherlockNoMad commented on torch/_ops.py:
```diff
@@ -377,6 +377,12 @@ def check_overloaded(arg):
+                    subclass_type = torch._subclasses.fake_tensor.FakeTensorMode  # type: ignore[assignment]
+                    handler = self.python_key_table[subclass_type]
+                    result = handler(arg.fake_mode, *args, **kwargs)  # type: ignore[attr-defined]
+                    return result
```
@zou3519 Does this change looks ok to you?

SherlockNoMad replied:
Here's the use case: during AOTI lowering, ppl wish to do numerical validation by running the lowered model eagerly. 

The model sometime has FakeTensor in it, for TBE weight. In such case, it will need the fixes here to make the eager run pass. It would produce a fake output. 

IIUC, you are suggesting we do a conditional setup for FakeMode. 

```
if model_has_fake_weights:
    with fake_mode_from_fake_weight: 
        model(args)
else:
    model(args)
```

Can we do this kind of check in our FakeTensor infra code? instead of asking user to set up the FakeMode?

zou3519 replied:
I'm convinced, thanks for the example. Can you leave a comment here saying something along the lines of "If people are using FakeTensor without a FakeTensorMode, pull the FakeTensorMode impl"?

SherlockNoMad replied:
sure.

zou3519 commented on torch/testing/_internal/hop_db.py:
```diff
@@ -75,6 +75,7 @@ def f2(x, y0, y1):
     "triton_kernel_wrapper_functional",
     "hints_wrapper",
     "foreach_map",
+    "aoti_call_delegate",
```
Can you add a test for this? It's important (for export tests too!) that this is tested

SherlockNoMad replied:
aoti_call_delegate accepts an AOTInductorEPModule, which is defined in fbcode only, so it's easier to test this in fbcode.
and I have switch PT2I tests to start consuming this HOP, see D68599208

zou3519 replied:
Fair enough, can you put a comment here explaining this?

zou3519 commented on torch/_higher_order_ops/aoti_call_delegate.py:
```diff
@@ -0,0 +1,72 @@
+# mypy: allow-untyped-defs
```
You've got pyre-strict but mypy: allow-untyped-defs. Let's remove mypy: allow-untyped-defs ?

SherlockNoMad replied:
done

zou3519 commented on torch/_higher_order_ops/aoti_call_delegate.py:
```diff
@@ -0,0 +1,72 @@
+    def __init__(self):
+        super().__init__("aoti_call_delegate")
+
+    def __call__(self, lowered_module, original_gm, weight_args, input_args):
```
Add a docstring?

SherlockNoMad replied:
done.

zou3519 reviewed (commented):
Would prefer not adding to the hop_db allowlist. We should also add a docstring to the aoti_call_delegate's `__call__`. Otherwise LGTM

angelayi commented on torch/_higher_order_ops/aoti_call_delegate.py:
```diff
@@ -0,0 +1,110 @@
+AOTI_LOWERED_MODULE = "AOTInductorEPModule"
+
+
+class AOTICallDelegate(HigherOrderOperator):
```
is it valuable to make this op more general? Like couldn't it also work for MTIA too?

SherlockNoMad committed:
Introduce aoti_call_delegate HOP (#145630)

Summary:

Previously, aoti compile node is represented as a kernel-less custom op in the exported program. The node was not eager runnable, which is a common practice for numerical validation during lowering. 

I introduce a new HOP to address this. 

The schema is following
```
aoti_call_delegate(lower_moduel: AOTInductorEPModule, original_gm: fx.GraphModule, weights: List[Tensor], inputs: List[Tensor])
```

There are a few problems exposed by HOP
- AOTI expects a FX graph with weights as getattr nodes, aka stateful graph. HOP expect graph_module arguments to be stateless. Export serializer also expect a stateless graph. Currently, to make AOTI happy, I am making `original_gm` stateful, and bypassing the serialization for `original_gm`. 
- As a result, the HOP is not re-traceable, as functionalization on stateful graph module argument will fail.

Test Plan: buck2 test 'fbcode//mode/opt' fbcode//deeplearning/aot_inductor/cpu/test:cpu_lowering_utils_test

Reviewed By: zhxchen17

Differential Revision: D68359391
